rfcs/51-identify-go-storage-error.md

```diff
@@ -0,0 +1,99 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-05-08
+---
+
+# AOS-51: Identify go-storage Error
+
+## Background
+
+Currently we use a function named `formatError` in generated code to turn SDK errors into our errors, as defined in [AOS-47]. However, sometimes we have used our errors before `formatError`, so our errors will also be formatted. Then it's hard for us to distinguish between non-SDK errors and our errors.
+
+Take `go-storage-s3` as an example:
+
+```go
+func formatError(err error) error {
+	e, ok := err.(awserr.RequestFailure)
+	if !ok {
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+	switch e.Code() {
+	// AWS SDK will use status code to generate awserr.Error, so "NotFound" should also be supported.
+	case "NoSuchKey", "NotFound":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	case "AccessDenied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, err)
+	default:
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+}
+```
+
+This is problematic since our `PairRequiredError` will also be turned into `ErrUnexpected`.
+
+## Proposal
+
+So I propose to add support for recognizing errors defined in `go-storage`, including:
+
+- Add a new type `ErrorCode` to replace sentinel errors.
+- Add a special `ErrorCode` `ErrGoStorage`
+- `ErrorCode` and every error `struct` should implement `Is(target error) bool { return target == ErrGoStorage}`
+
+## Rationale
+
+We cannot add methods to sentinel errors, so we need to replace them.
+
+### Alternative
+
+Add an interface and use type cast to judge.
+
+```go
+type ErrGoStorage interface {
```

Comment:
How define an internal interface like following:

```go
type AosError interface {
	isAosError()
}
```

We will have code like the following:

```go
type Error string

func (e Error) Error() string {
	return string(e)
}

func (e Error) isAosError() {}

type AosError interface {
	isAosError()
}
```

We only need to add an empty function for our errors.

Here are benchmark tests:

```go
import (
	"errors"
	"testing"
)

type Error string

var specialError = errors.New("x")

func (e Error) Error() string {
	return string(e)
}

func (e Error) Is(err error) bool {
	return true
}

func (e Error) isAosError() {}

type AosError interface {
	isAosError()
}

func BenchmarkInternalInterface(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_ = e.(AosError)
	}
}

func BenchmarkErrorsIs(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_ = errors.Is(e, specialError)
	}
}

Resultes:

```
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkInternalInterface
BenchmarkInternalInterface-8   	176883178	         6.845 ns/op
BenchmarkErrorsIs
BenchmarkErrorsIs-8            	78588775	        13.68 ns/op
PASS
```

Use our own interface is a bit faster.

Reply:
Benchmark is convincing! I will also try to use benchmark to compare alternatives in the future :p

Reply:
If `isAosError` is an unexported method, what if a service creates its error `struct`? Then it cannot implement `AosError`

Reply:
They should define new error by using `services.Error` which already implement it.

Reply:
A workaround is to embed `ErrorCode` as an anonymous field into error `struct`s... (The hidden `Err` approach)

Btw, I was just considering making `ErrorCode` private... Just like `errors.New()` returns private `stringError` 

So two choices:
1. private `isAosError` + public `ErrorCode` + embed `ErrorCode` in `struct`
2. public `IsAosError` + private (or public) `errorCode` + no embed

Reply:
> If `isAosError` is an unexported method, what if a service creates its error `struct`? Then it cannot implement `AosError`.

It OK to implement an interface even there is an unexported method. I make this function private is to restrict user's ability.

Reply:
Let's move the `AosError` definition to another package:

```go
func (e Error) Error() string {
	return string(e)
}

func (e Error) Is(err error) bool {
	return true
}

func (e Error) isAosError() {}

func BenchmarkInternalInterface(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_, _ = e.(types.AosError)
	}
}
```

It still works.

Reply:
Take a look at `_, ok := e.(a.AosError)`, `ok` is `false`

Reply:
Oh, I made a mistake.

Reply:
```go
var e (a.AosError)
e = Error("xxxx")
```
can find problem more quickly

Reply:
OK, let's export `IsAosError`, but with comment like:

```go
// IsAosError should only be implemented by go-storage & go-service.
// We depends this function and related AosError interface to detect error.
```